mattermost/mattermost-mobile
# MM-45745 - Calls channel info screen options

#### Summary
- Implement the start/join call and enable/disable calls option in the channel_info component.
- Refactor the join call logic to better handle and report connection errors back to the user. Now we have some friendly feedback. I'm hoping to improve this further to give better debugability to mobile.

#### Testing
Here are the test cases. @Phrynobatrachus you might want to start tracking these. I hope we can automate these eventually because it's very easy for regressions to creep in (e.g., the enable/disable silently broke).

Pressing disable/enable option in any of the cases when disable/enable button is shown should flip disable/enable option.

Plugin enabled/disabled:
- plugin disabled before loading app: enable/disabled option will get "(Not available)" after pressing, and pressing gives alert.
- plugin disabled after loading: both options will show, will get "(Not available)" after pressing, and pressing gives alert.

System Console Setting: Enable on specific channels = false & default enable on all channels = false:
- on a brand new channel, as regular user, do not show disable/enable option, does not show start call.
- on a brand new channel, as channel admin, do not show disable/enable option, does not show start call.
- on a brand new channel, as system admin, shows disable/enable option, does not show start call. disable/enable clickable and changes state accordingly.

System Console Setting: Enable on specific channels & default not enable on all channels:
- on a brand new channel, as regular user, do not show disable/enable option, does not show start call.
- on a brand new channel, as channel admin, shows disable/enable option, does not show start call. disable/enable clickable and changes state accordingly.
- on a brand new channel, as system admin, shows disable/enable option, does not show start call. disable/enable clickable and changes state accordingly.
  
System Console Setting: Enable on specific channels & default enabled on all channels:
- As regular user not channel admin, do not show disable/enable option. As channel admin, show disable/enable option. As system admin, show disable/enable option. All can start call.
- server with default enabled and channel explicitly disabled: shows disabled, and start call button not shown. As regular user not channel admin, do not show disable/enable option. As channel admin, show disable/enable option. As system admin, show disable/enable option.

Discovered an edge case:
- System Console Setting: Enable on specific channels & default enabled on all channels: As regular user loggin in but not in channelA, disable calls in channelA, join channelA, "Start Call" option will be visable, but pressing it does nothing. Will not show "Start Call" option after user reconnects and pulls down updated list of call channels enabled.
 
Cases for start/joining a call:
- start in a channel without user currently in a call. Channel info option should say "start". No confirmation dialog. User should be able to start the call.
- join in a channel without the user currently in a call. Channel info option should say "Join Ongoing Call". No confirmation dialog. User should be able to join.
- start in a channel without ongoing call but user is currently in a call. Channel info option should say "start". Should show an alert with the proper channel names for the channel leaving and channel joining. Should say "start" in the message description. User should be able to join.
- join in a channel with an ongoing call but user is currently in a call. Channel info option should say "join." Should show an alert with the proper channel names for the channel leaving and channel joining. User should be able to join.
- start in a channel without ongoing call but user is currently in a call on a different server. Channel info option should say "start". Should show an alert with the proper channel names for the channel leaving and channel joining. Should say "start" in the message description. User should be able to join.
- join in a channel with an ongoing call but user is currently in a call on a different server. Channel info option should say "join." Should show an alert with the proper channel names for the channel leaving and channel joining. User should be able to join.

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-45745

#### Checklist
- [x] Includes text changes and localization file updates
- [x] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
- Android: 12, Pixel 6
- iOS: 15.3.1, iPhone 7 plus

#### Screenshots
When there are no options (calls disabled, regular user), same as before this PR:
<img width="221" alt="image" src="https://user-images.githubusercontent.com/1490756/180494153-ba3e9ddb-e4ac-4a16-a222-7c54693d27ff.png">

Various combinations:
<img width="224" alt="image" src="https://user-images.githubusercontent.com/1490756/180494218-fc8592f4-5211-4880-af40-a5bf3c6c640c.png">
<img width="233" alt="image" src="https://user-images.githubusercontent.com/1490756/180494291-453f346c-5fa6-40ae-9bb9-b3ca26937925.png">
<img width="235" alt="image" src="https://user-images.githubusercontent.com/1490756/180494323-8d9a432e-c682-43dd-8221-858200f3faa7.png">
<img width="228" alt="image" src="https://user-images.githubusercontent.com/1490756/180494400-8f977a2c-03aa-492f-b88a-028b74d770d8.png">


#### Release Note
```release-note
Calls: Start/Join Call and Enable/Disable Calls options added to the channel info screen.
```
